### PR TITLE
Implement parallel downloads for COM scenarios

### PR DESCRIPTION
**Pull request by lechacon**
This changes allows multiple installer downloads to occur in parallel (and parallel to installation) on COM scenarios.

Modified the context orchestrator to have multiple queues for items to execute, each of which will handle a type of command (download & install). Items to execute will move to each queue according to the next command to execute for the item. Each queue handles execution of its items.
Adding a queue item in the orchestrator will add it to the queue corresponding to its first command. Once it is executed, it is moved to the queue for the following command and so on. Each queue knows how many commands of its type can run at the same time (n for download, 1 for install) and how many are currently running. When an item is added, if it has capacity, it starts a thread to run it. And when a thread finishes running a command (after sending it to the orchestrator to re-queue), it runs the next item.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/1588)

### CODE REVIEW

**Comment by JohnMcPMS** [comment]
@msftbot, leave the PR open.

[Maybe that works?]

File: src/AppInstallerRepositoryCore/Microsoft/SQLiteIndex.h [hunk 1]
```diff
@@ -139,5 +139,6 @@ namespace AppInstaller::Repository::Microsoft
         SQLite::Connection m_dbconn;
         Schema::Version m_version;
         std::unique_ptr<Schema::ISQLiteIndex> m_interface;
+        std::unique_ptr<std::mutex> m_interfaceLock = std::make_unique<std::mutex>();
```
[the remark this answers is not in the document]
Comment (lechacon) [reply src/AppInstallerRepositoryCore/Microsoft/SQLiteIndex.h:142]: We need to be able to move the SQLiteIndex (for example, in the Create* functions for it) but a mutex cannot be moved, so it cannot be a member directly and I'm just keeping a pointer to it.

File: src/AppInstallerCLICore/ContextOrchestrator.cpp [hunk 2]
```diff
@@ -21,89 +21,159 @@ namespace AppInstaller::CLI::Execution
         ::AppInstaller::ProgressCallback progress;
         std::shared_ptr<::AppInstaller::Repository::ISource> installingSource = ::AppInstaller::Repository::OpenPredefinedSource(::AppInstaller::Repository::PredefinedSource::Installing, progress);
         m_installingWriteableSource = std::dynamic_pointer_cast<::AppInstaller::Repository::IMutablePackageSource>(installingSource);
+
+        AddCommandQueue(COMDownloadCommand::CommandName, 5);
```
[the remark this answers is not in the document]
Comment (lechacon) [reply src/AppInstallerCLICore/ContextOrchestrator.cpp]: Changed to use `hardware_concurrency`, and default to 1 if we can't determine it. Not sure if we should add an upper limit even if hardware supports many threads, or use multiple even if it doesn't.

File: src/AppInstallerCLICore/ContextOrchestrator.cpp [hunk 3]
```diff
@@ -21,89 +21,159 @@ namespace AppInstaller::CLI::Execution
...
-    void ContextOrchestrator::EnqueueAndRunItem(std::shared_ptr<OrchestratorQueueItem> item)
-    {
-        EnqueueItem(item);
-
-        std::thread runnerThread(&ContextOrchestrator::RunItems, this);
-        runnerThread.detach();
+        {
+            // Start a runner for this item if there is capacity for it
+            std::lock_guard<std::mutex> lockQueue{ m_threadsLock };
+            if (m_runningThreads < m_allowedThreads)
```
[the remark this answers is not in the document]
Comment (lechacon) [reply src/AppInstallerCLICore/ContextOrchestrator.cpp]: I changed the implementation to use Windows Thread Pools.

File: src/AppInstallerRepositoryCore/Microsoft/SQLiteIndex.cpp [hunk 4]
```diff
@@ -143,6 +143,7 @@ namespace AppInstaller::Repository::Microsoft
     {
         AICLI_LOG(Repo, Verbose, << "Adding manifest for [" << manifest.Id << ", " << manifest.Version << "] at relative path [" << relativePath << "]");
 
+        std::lock_guard<std::mutex> lockInterface{ *m_interfaceLock };
```
[the remark this answers is not in the document]
Comment (lechacon) [reply src/AppInstallerRepositoryCore/Microsoft/SQLiteIndex.cpp:154]: Done.

File: .github/actions/spelling/allow.txt [hunk 5]
```diff
@@ -533,6 +533,7 @@ testsettingname
 TEXTFORMAT
 TEXTINCLUDE
 there're
```
Comment (lechacon) [inline .github/actions/spelling/allow.txt:535]: Do we really need this one? It seems weird to me; I think we could change the one use of it to `there are`

Comment (JohnMcPMS) [reply .github/actions/spelling/allow.txt:535]: I agree, although apparently it is somewhat acceptable I have never seen it before.

File: src/AppInstallerCLICore/ContextOrchestrator.cpp [hunk 6]
```diff
@@ -10,6 +10,21 @@
 
 namespace AppInstaller::CLI::Execution
 {
+    namespace
+    {
+        // Callback function used by worker threads in the queue.
+        // context must be a pointer to a queue item.
+        void OrchestratorQueueWorkCallback(PTP_CALLBACK_INSTANCE, PVOID context, PTP_WORK)
+        {
+            auto queueItem = static_cast<OrchestratorQueueItem*>(context);
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/ContextOrchestrator.cpp:19]: I prefer `reinterpret_cast` in these cases to make it clearer that this is not type safe.

File: src/AppInstallerCLICore/ContextOrchestrator.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (JohnMcPMS) [reply src/AppInstallerCLICore/ContextOrchestrator.cpp]: I think that an upper limit is reasonable; at some point we would saturate the network and the goal is that at least one should be downloaded by the time the install finishes.  Maybe 3?

File: src/AppInstallerCLICore/ContextOrchestrator.cpp [hunk 7]
```diff
@@ -21,90 +36,192 @@ namespace AppInstaller::CLI::Execution
...
         }
+    }
 
-        // Running state must be set inside the queueLock so that multiple threads don't try to run the same item.
-        item->SetState(OrchestratorQueueItemState::Running);
-        return item;
+    OrchestratorQueue::OrchestratorQueue(std::string_view commandName, UINT32 allowedThreads) :
+        m_commandName(commandName), m_allowedThreads(allowedThreads), m_threadPool(CreateThreadpool(nullptr)), m_threadPoolCleanupGroup(CreateThreadpoolCleanupGroup())
+    {
+        THROW_LAST_ERROR_IF_NULL(m_threadPool);
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/ContextOrchestrator.cpp:164]: If they both fail, this would return the error from the cleanup group.

**Review by JohnMcPMS — APPROVED** [review]

**Pull request merged**